src/internal/packager/git/refs.go

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package git contains functions for interacting with git repositories
+package git
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/defenseunicorns/zarf/src/pkg/message"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+)
+
+// removeLocalBranchRefs removes all refs that are local branches
+// It returns a slice of references deleted
+func (g *Git) removeLocalBranchRefs() ([]*plumbing.Reference, error) {
+	return g.removeReferences(
+		func(ref *plumbing.Reference) bool {
+			return ref.Name().IsBranch()
+		},
+	)
+}
+
+// removeOnlineRemoteRefs removes all refs pointing to the online-upstream
+// It returns a slice of references deleted
+func (g *Git) removeOnlineRemoteRefs() ([]*plumbing.Reference, error) {
+	return g.removeReferences(
+		func(ref *plumbing.Reference) bool {
+			return strings.HasPrefix(ref.Name().String(), onlineRemoteRefPrefix)
+		},
+	)
+}
+
+// removeHeadCopies removes any refs that aren't HEAD but have the same hash
+// It returns a slice of references deleted
+func (g *Git) removeHeadCopies() ([]*plumbing.Reference, error) {
+	message.Debugf("git.removeHeadCopies()")
```

Comment:
I'm a bit torn on this one myself, I think it is fine to reduce these if we start including stack trace with debug entries on failure--otherwise we should probably just do them everywhere to be consistent. Before we didn't have the concept of a debug log so these were more necessary. Now I think we could probably just tweak our fatal behavior and strip out all the function name debugs. They're also less valuable now that we're using this pattern.

Reply:
My instinct is that I (surprisingly) prefer keep these (and adding them where they are missing.

I think the noise these logs leave isn't that crazy but they could be really useful if we have something unexpected happening that isn't actually causing an error/failure or if we have some weird infinite loop / threading death-trap that isn't properly being handled.

Reply:
I have actually not really used these myself.  If our debug/error messages are descriptive/unique enough then it is not hard to find out where they are coming from IMO and then you can reconstruct the trace through various functions from there (this is generally how I have done debugging in the past).  I would be for removing the function name debug logs and then doing stack traces on fatals instead.

Reply:
quick ref, looking at calling the trace

![Screenshot 2022-11-28 at 6 33 22 PM](https://user-images.githubusercontent.com/882485/204409377-b8479783-9104-4da5-9c81-01cac9b3f9a0.png)

Reply:
(Also I think that we could make better use of "trace" logging which has no difference from debug logging right now if we wanted to.  It seems to have been removed but there was an annoying debug message that would always print the full JSON state of a pod and fill the terminal with that message making it hard to debug other things)

Reply:
Stacktrace on fatal
![Screenshot 2022-11-30 at 4 15 15 AM](https://user-images.githubusercontent.com/882485/204775399-20e7755f-171b-4125-81ce-7ca58a2a34ce.png)

Reply:
But there seems to be a problem with the split debug logging, where the debug logs are not written properly to the log file in non-debug mode.

Reply:
> But there seems to be a problem with the split debug logging, where the debug logs are not written properly to the log file in non-debug mode.

Pushed a fix, we were shadowing the log file pointer with a `:=` so once we left the local scope the rest of the application was still seeing the file location as a nil pointer.